execution/executor/src/components/apply_chunk_output.rs

```diff
@@ -0,0 +1,272 @@
+        } = state_cache;
+
+        // Apply write sets to account states in the AccountCache, resulting in new account states.
+        let account_states = to_keep
+            .iter()
+            .map(|(t, o)| process_write_set(t, &mut accounts, o.write_set().clone()))
+            .collect::<Result<Vec<_>>>()?;
+        let account_blobs = account_states
+            .par_iter()
+            .with_min_len(100)
+            .map(|account_to_state| {
+                account_to_state
+                    .iter()
+                    .map(|(addr, state)| Ok((*addr, AccountStateBlob::try_from(state)?)))
+                    .collect::<Result<HashMap<_, _>>>()
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // apply new account states to the base state tree, resulting in updated state tree.
+        let (roots_with_node_hashes, result_state) = frozen_base
+            .serial_update(
+                Self::account_blobs_to_smt_updates(&account_blobs),
+                &ProofReader::new(proofs),
+            )
+            .map_err(|e| anyhow!("Failed to update state tree. err: {:?}", e))?;
+        let result_state = result_state.unfreeze();
+
+        // Get the updated validator set from updated account state.
+        let next_epoch_state = if new_epoch {
+            Some(Self::parse_validator_set(&accounts)?)
+        } else {
+            None
+        };
+
+        Ok((
+            account_blobs,
+            roots_with_node_hashes,
+            result_state,
+            next_epoch_state,
+        ))
+    }
+
+    fn account_blobs_to_smt_updates(
+        account_blobs: &[HashMap<AccountAddress, AccountStateBlob>],
+    ) -> Vec<Vec<(HashValue, &AccountStateBlob)>> {
+        account_blobs
+            .iter()
+            .map(|m| {
+                m.iter()
+                    .map(|(account, blob)| (HashAccountAddress::hash(account), blob))
+                    .collect::<Vec<_>>()
+            })
+            .collect()
+    }
+
+    fn parse_validator_set(accounts: &HashMap<AccountAddress, AccountState>) -> Result<EpochState> {
+        let validator_set = accounts
+            .get(&on_chain_config::config_address())
+            .map(|state| {
+                state
+                    .get_validator_set()?
+                    .ok_or_else(|| anyhow!("ValidatorSet does not exist"))
+            })
+            .ok_or_else(|| anyhow!("ValidatorSet account does not exist"))??;
+        let configuration = accounts
+            .get(&on_chain_config::config_address())
+            .map(|state| {
+                state
+                    .get_configuration_resource()?
+                    .ok_or_else(|| anyhow!("Configuration does not exist"))
+            })
+            .ok_or_else(|| anyhow!("Association account does not exist"))??;
+
+        Ok(EpochState {
+            epoch: configuration.epoch(),
+            verifier: (&validator_set).into(),
+        })
+    }
+
+    fn assemble_ledger_diff(
```

Comment:
the name seems confusing, you mean compute_txn_data?

Reply:
transaction_data is less expressive in my opinion.. (maybe call the type ExecutedTransaction instead of TransactionData?)

Reply:
what's the `diff` referred to? txn to be committed?

Reply:
all the content (txns, txn outputs, state roots) needed to advance the ledger to a newer version